omx/gstomxcamera.c

```diff
@@ -0,0 +1,1214 @@
+        ("Cannot acquire output buffer from pool"));
+    return flow_ret;
+  }
+release_error:
+  {
+    GST_ELEMENT_ERROR (self, LIBRARY, SETTINGS, (NULL),
+        ("Failed to relase output buffer to component: %s (0x%08x)",
+            gst_omx_error_to_string (err), err));
+    gst_pad_push_event (GST_BASE_SRC_PAD (self), gst_event_new_eos ());
+    self->started = FALSE;
+    return GST_FLOW_ERROR;
+  }
+}
+
+
+static gboolean
+gst_omx_camera_component_init (GstOMXCamera * self, GList * buffers)
+{
+  GST_DEBUG_OBJECT (self, "Enabling buffers");
+  if (gst_omx_port_set_enabled (self->outport, TRUE) != OMX_ErrorNone)
+    return FALSE;
+
+  if (gst_omx_port_wait_enabled (self->outport,
+          1 * GST_SECOND) != OMX_ErrorNone)
+    return FALSE;
+
+  GST_DEBUG_OBJECT (self, "Changing state to Idle");
+  if (gst_omx_component_set_state (self->comp, OMX_StateIdle) != OMX_ErrorNone)
+    return FALSE;
+
+  /* Need to allocate buffers to reach Idle state */
+  if (!buffers) {
+    if (gst_omx_port_allocate_buffers (self->outport) != OMX_ErrorNone)
+      return FALSE;
+  } else {
+    if (gst_omx_port_use_buffers (self->outport, buffers) != OMX_ErrorNone) {
+      return FALSE;
+    }
+  }
+
+  if (gst_omx_component_get_state (self->comp,
+          GST_CLOCK_TIME_NONE) != OMX_StateIdle)
+    return FALSE;
+
+  GST_DEBUG_OBJECT (self, "Changing state to Executing");
+  if (gst_omx_component_set_state (self->comp,
+          OMX_StateExecuting) != OMX_ErrorNone)
+    return FALSE;
+
+  if (gst_omx_component_get_state (self->comp,
+          GST_CLOCK_TIME_NONE) != OMX_StateExecuting)
+    return FALSE;
+
+  gst_omx_port_set_flushing (self->outport, 5 * GST_SECOND, FALSE);
+
+  if (gst_omx_port_populate (self->outport) != OMX_ErrorNone)
+    return FALSE;
+  if (gst_omx_port_mark_reconfigured (self->outport) != OMX_ErrorNone)
+    return FALSE;
+
+  /* Allocate src buffer pool buffers */
+  GST_OMX_BUFFER_POOL (self->outpool)->allocating = TRUE;
+  if (!gst_buffer_pool_set_active (self->outpool, TRUE))
+    return FALSE;
+  GST_OMX_BUFFER_POOL (self->outpool)->allocating = FALSE;
+  GST_OMX_BUFFER_POOL (self->outpool)->deactivated = FALSE;
+
+
+  return TRUE;
+}
+
+
+static GstFlowReturn
+gst_omx_camera_create (GstPushSrc * src, GstBuffer ** buf)
+{
+  GstOMXCamera *self = GST_OMX_CAMERA (src);
+  GstFlowReturn ret;
+  GstClock *clock;
+  GstClockTime abs_time, base_time, timestamp;
+
+  /* timestamps, LOCK to get clock and base time. */
+  GST_OBJECT_LOCK (self);
+  if ((clock = GST_ELEMENT_CLOCK (self))) {
+    /* we have a clock, get base time and ref clock */
+    base_time = GST_ELEMENT (self)->base_time;
+    abs_time = gst_clock_get_time (clock);
+  } else {
+    /* no clock, can't set timestamps */
+    base_time = GST_CLOCK_TIME_NONE;
+    abs_time = GST_CLOCK_TIME_NONE;
+  }
+  GST_OBJECT_UNLOCK (self);
+
+  if (!self->started) {
+    if (!gst_omx_camera_component_init (self, NULL)) {
```

Comment:
Can we do this in the `start`?

Reply:
@cgomezv  Calling gst_omx_camera_component_init inside the start causes plugin/component failures. This happens because the component_init gets in charge of moving the component from loaded to the executing state, so calling it during the start makes the set_format function to fail. This occurs because set_format needs the component in loaded state in order to configure the omx ports, so it will send the component back to loaded state and the effort of doing the component_init in the start becomes useless.